Add missing getters to navigation servers

nickyfoo opened the pull request:
Closes #84495 

Add virtual functions and bind to navigation servers
Implement getters
Add documentation
<!--
Please target the `master` branch in priority.
PRs can target `3.x` if the same change was done in `master`, or is not relevant there.

Relevant fixes are cherry-picked for stable branches as needed by maintainers.
You can mention in the description if the change is compatible with `3.x`.

To speed up the contribution process and avoid CI errors, please set up pre-commit hooks locally:
https://docs.godotengine.org/en/latest/contributing/development/code_style_guidelines.html
-->

nickyfoo commented on modules/navigation/godot_navigation_server.h:
```diff
@@ -96,6 +96,8 @@ class GodotNavigationServer : public NavigationServer3D {
 	int pm_edge_connection_count = 0;
 	int pm_edge_free_count = 0;
 
+	const Vector<Vector3> dummy_empty_vertices;
```
Addition of this so that `obstacle_get_vertices` is able to return by const ref (to avoid returning a dangling ref)

smix8 replied:
I think we should remove the ref part same as with 2D, the servers usually do not return anything by ref.

smix8 replied:
Can now be removed imo.

nickyfoo commented on modules/navigation/godot_navigation_server_2d.cpp:
```diff
@@ -141,6 +141,13 @@ static Transform3D trf2_to_trf3(const Transform2D &d) {
 	return Transform3D(b, o);
 }
 
+static Transform2D trf3_to_trf2(const Transform3D &d) {
```
When tested locally with the following GDScript, I get output that seems to be within epsilon.
```
extends Node2D


var region: RID = NavigationServer2D.region_create()
var t = Transform2D()
var rot = -0.5 # The rotation to apply.

# Called when the node enters the scene tree for the first time.
func _ready():
	t.origin = Vector2(350,150)
	t.x.x = cos(rot)
	t.y.y = cos(rot)
	t.x.y = sin(rot)
	t.y.x = -sin(rot)
	# Scale
	t.x *= 3
	t.y *= 3
	print(t)
	NavigationServer2D.region_set_transform(region, t)
	pass # Replace with function body.


# Called every frame. 'delta' is the elapsed time since the previous frame.
func _process(delta):
	print(t, NavigationServer2D.region_get_transform(region))
	pass

OUTPUT: 
[X: (2.632748, -1.438277), Y: (1.438277, 2.632748), O: (350, 150)]
[X: (2.632751, -1.438278), Y: (1.438263, 2.632751), O: (350, 150)]
```

smix8 replied:
That is fine, a small difference is to be expected from converting back and forth between 2D and 3D.

nickyfoo commented on servers/navigation_server_2d.h:
```diff
@@ -262,10 +276,15 @@ class NavigationServer2D : public Object {
 	virtual void obstacle_set_position(RID p_obstacle, Vector2 p_position) = 0;
+	virtual Vector2 obstacle_get_position(RID p_obstacle) const = 0;
 	virtual void obstacle_set_vertices(RID p_obstacle, const Vector<Vector2> &p_vertices) = 0;
+	virtual Vector<Vector2> obstacle_get_vertices(RID p_obstacle) const = 0;
```
This signature is not a const ref because it is difficult to implement (since internally the 2D server is a wrapper over the 3D server, to return this by ref would have to dynamically allocate the result of a conversion from `Vector<Vector3>` to `Vector<Vector2>`, otherwise it would be a dangling ref.

I'm not sure how much the const ref would help/how it translates for the engine though.

smix8 replied:
That is ok, I don't think we should use ref in that case anyway, the servers normally do not return refs directly to users.

smix8 commented on modules/navigation/godot_navigation_server.cpp:
```diff
@@ -899,37 +997,79 @@ COMMAND_2(obstacle_set_radius, RID, p_obstacle, real_t, p_radius) {
 
+Vector<Vector3> GodotNavigationServer::obstacle_get_vertices(RID p_obstacle) const {
+	NavObstacle *obstacle = obstacle_owner.get_or_null(p_obstacle);
+	ERR_FAIL_NULL_V(obstacle, dummy_empty_vertices);
```
I think with the ref removed this `dummy_empty_vertices` can now be `Vector<Vector3>()` as it is more familiar and commonly used across the engine in similar cases.

nickyfoo replied:
Yup, I agree, thanks for the catch!

smix8 commented on modules/navigation/godot_navigation_server_2d.h:
```diff
@@ -51,6 +51,8 @@ class GodotNavigationServer2D : public NavigationServer2D {
 	NavMeshGenerator2D *navmesh_generator_2d = nullptr;
 #endif // CLIPPER2_ENABLED
 
+	const Vector<Vector2> dummy_empty_vertices;
```
Can be removed? I dont see any uses of this in the 2D server.

nickyfoo replied:
Ah yeah, of course, it slipped my mind to remove them after removing the ref. Thanks!

smix8 reviewed (commented):
Looks good.

As mentioned now that the ref is gone I would remove the dummy and replace the single use in the obstacle vertices error macro with a normal Vector<Vector3>(). It is just more common this way.

smix8 reviewed (approved):
Thanks.

nickyfoo committed:
Add getters to navigation servers

Add virtual functions and bind to navigation servers
Implement getters
Add documentation

YuriSizov commented:
Rebased on your behalf.

YuriSizov commented:
Thanks, and congrats on your first merged Godot contribution!
